pkg/hive/cell/decorators.go

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package cell
+
+import (
+	"github.com/cilium/cilium/pkg/hive/internal"
+)
+
+// Decorate takes a decorator function and a set of cells and returns
+// a decorator cell.
+//
+// A decorator function is a function that takes as arguments objects
+// in the hive and returns one or more augmented objects. The cells wrapped
+// with a decorator will be provided the returned augmented objects.
+//
+// Example:
+//
+//	cell.Decorate(
+//		func(e Example) Example {
+//			return e.WithMoreMagic()
+//		},
+//		cell.Invoke(func(e Example) {
+//			// e now has more magic
+//		},
+//	)
+func Decorate(dtor any, cells ...Cell) Cell {
+	return &decorator{
+		decorator: dtor,
+		cells:     cells,
+	}
+}
+
+type decorator struct {
+	decorator any
+	cells     []Cell
+}
+
+func (d *decorator) Apply(c container) error {
+	scope := c.Scope("(decorate)")
```

Comment:
To me it seemed to make more sense to wrap things with a decorator rather than having the decorator be associated with a module and it’s scope. That’s why I ended up with this. Name of the scope could be better.